### PR TITLE
cosTheta ReferenceLine Smoother(with constraint)

### PR DESCRIPTION
**Pull request by JasonZhou404**

### CODE REVIEW

File: modules/planning/conf/planning.conf [hunk 1]
```diff
@@ -10,3 +10,8 @@
 --enable_multi_thread_in_dp_poly_path
 --enable_multi_thread_in_dp_st_graph
 --default_cruise_speed=20.00
+# --smoother_config_filename=modules/planning/conf/spiral_smoother_config.pb.txt
+# --smoother_config_filename=modules/planning/conf/cosTheta_smoother_config.pb.txt
+# --enable_reference_line_stitching=false
+
```
Comment (lianglia-apollo) [inline modules/planning/conf/planning.conf]: remove the empty lines

File: modules/planning/proto/reference_line_smoother_config.proto [hunk 2]
```diff
@@ -42,6 +42,29 @@ message SpiralSmootherConfig {
...
+  // Relax constraint on end condition
+  optional double relax = 5 [default = 0.2];
+
+  // kappa filter
+  optional double kappa_filter = 6 [default = 1.0e9];
+
+  // dkappa filter
+  optional double dkappa_filter = 7 [default = 1.0e9];
+
+  }
```
Comment (lianglia-apollo) [inline modules/planning/proto/reference_line_smoother_config.proto]: Please fix the indentation of the braces.

File: modules/planning/reference_line/cosTheta_reference_line_smoother.h [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+  double zero_x_ = 0.0;
+
+  double zero_y_ = 0.0;
+};
+
+}  // namespace planning
+}  // namespace apollo
+
+#endif  // MODULES_PLANNING_REFERENCE_LINE_cosTheta_REFERENCE_LINE_SMOOTHER_H_
```
Comment (lianglia-apollo) [inline modules/planning/reference_line/cosTheta_reference_line_smoother.h]: cosTheta --> COSTHETA

File: modules/planning/reference_line/cosTheta_problem_interface.cc [hunk 4]
```diff
@@ -0,0 +1,2043 @@
...
+  // for (std::size_t i = 0; i < num_of_points_; i++) {
+  //   lambda[i] = 0.0;
+  // }
+
+  return true;
+}
+
+bool cosThetaProbleminterface::eval_f(int n, const double* x, bool new_x,
+                                      double& obj_value) {
+  assert(std::size_t(n) == num_of_variables_);
```
[the remark this answers is not in the document]
Comment (lianglia-apollo) [reply modules/planning/reference_line/cosTheta_problem_interface.cc]: Also, please use static_cast<size_t>

**Comment by JasonZhou404** [comment]
Please review again, I just push another commit to fix the problem you mentioned

File: modules/planning/reference_line/cosTheta_reference_line_smoother.cc [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+        << (interpl_end_timestamp - opt_end_timestamp) * 1000 << " ms.";
+
+  return status == Ipopt::Solve_Succeeded ||
+         status == Ipopt::Solved_To_Acceptable_Level;
+}
+
+void CosThetaReferenceLineSmoother::quintic_hermite_point(
+    const double t, const common::PathPoint front_point,
+    const common::PathPoint back_point, double* quintic_hermite_point_info) {
+  double q1 = -6 * t * t * t * t * t + 15 * t * t * t * t - 10 * t * t * t + 1;
```
Comment (startcode) [inline modules/planning/reference_line/cosTheta_reference_line_smoother.cc]: calculate pow(t) first;

File: modules/planning/reference_line/cosTheta_reference_line_smoother.cc [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+    }
+    est1 = multiplier * (endsum + 2 * asum + 4 * bsum);
+  }
+  return est1;
+}
+
+double CosThetaReferenceLineSmoother::quintic_hermite_s(
+    const double t, common::PathPoint front_point,
+    common::PathPoint back_point) {
+  double q1 = -30 * t * t * t * t + 60 * t * t * t - 30 * t * t;
```
Comment (startcode) [inline modules/planning/reference_line/cosTheta_reference_line_smoother.cc]: pre-calculate pow(t)

File: modules/planning/reference_line/cosTheta_reference_line_smoother.cc [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+                       kappa, dkappa));
+  }
+
+  if (ref_points.size() < 2) {
+    AERROR << "Fail to generate smoothed reference line.";
+    return false;
+  }
+  *smoothed_reference_line = ReferenceLine(ref_points);
+  const double end_timestamp = Clock::NowInSeconds();
+  AINFO << "cosTheta reference line loading time: "
```
Comment (startcode) [inline modules/planning/reference_line/cosTheta_reference_line_smoother.cc]: use ADEBUG for frequently recurring logs when checkin code

File: modules/planning/reference_line/cosTheta_reference_line_smoother.cc [hunk 8]
```diff
@@ -0,0 +1,512 @@
...
+        << (end_timestamp - load_reference_timestamp) * 1000 << " ms.";
+  ADEBUG << "cosTheta reference line smoother time: "
+         << (end_timestamp - start_timestamp) * 1000 << " ms.";
+  AINFO << "cosTheta reference line smoother time: "
```
Comment (startcode) [inline modules/planning/reference_line/cosTheta_reference_line_smoother.cc]: use ADEBUG for frequently recurring logs when checkin code

File: modules/planning/reference_line/cosTheta_reference_line_smoother.cc [hunk 9]
```diff
@@ -0,0 +1,512 @@
...
+    // Retrieve some statistics about the solve
+    Ipopt::Index iter_count = app->Statistics()->IterationCount();
+    std::cout << "*** The problem solved in " << iter_count << " iterations!"
+              << std::endl;
+
+    Ipopt::Number final_obj = app->Statistics()->FinalObjective();
+    std::cout << "*** The final value of the objective function is "
+              << final_obj << "." << std::endl;
+  } else {
+    std::cout << "Return status: " << int(status) << std::endl;
```
Comment (startcode) [inline modules/planning/reference_line/cosTheta_reference_line_smoother.cc]: use AERROR or AINFO

File: modules/planning/reference_line/cosTheta_reference_line_smoother.cc [hunk 10]
```diff
@@ -0,0 +1,512 @@
...
+    return static_cast<int>(status);
+  }
+
+  status = app->OptimizeTNLP(problem);
+
+  if (status == Ipopt::Solve_Succeeded ||
+      status == Ipopt::Solved_To_Acceptable_Level) {
+    // Retrieve some statistics about the solve
+    Ipopt::Index iter_count = app->Statistics()->IterationCount();
+    std::cout << "*** The problem solved in " << iter_count << " iterations!"
```
Comment (startcode) [inline modules/planning/reference_line/cosTheta_reference_line_smoother.cc]: use AERROR or AINFO

File: modules/planning/reference_line/cosTheta_reference_line_smoother.cc [hunk 11]
```diff
@@ -0,0 +1,512 @@
...
+              << std::endl;
+
+    Ipopt::Number final_obj = app->Statistics()->FinalObjective();
+    std::cout << "*** The final value of the objective function is "
```
Comment (startcode) [inline modules/planning/reference_line/cosTheta_reference_line_smoother.cc]: use AERROR or AINFO

File: modules/planning/reference_line/cosTheta_problem_interface.cc [hunk 12]
```diff
@@ -0,0 +1,592 @@
...
+          obj_factor * (-weight_cos_included_angle_) *
+          ((q6 * q14) / (2 * q1 * std::sqrt(q1) * std::sqrt(q2)) -
+           1 / (std::sqrt(q1) * std::sqrt(q2)) +
+           (q9 * q11) / (2 * std::sqrt(q1) * q2 * std::sqrt(q2)) -
+           (q6 * q9 * q4) / (4 * q1 * std::sqrt(q1) * q2 * std::sqrt(q2)));
+
+      values[idx_map_[std::make_pair(topleft + 2, topleft + 2)]] +=
+          obj_factor * (-weight_cos_included_angle_) *
+          ((q5 * q7) / (q1 * std::sqrt(q1) * std::sqrt(q2)) -
+           q4 / (std::sqrt(q1) * q2 * std::sqrt(q2)) -
```
Comment (lianglia-apollo) [inline modules/planning/reference_line/cosTheta_problem_interface.cc]: I recommend record the results for sqrt(xxx) and reuse.

**Comment by JasonZhou404** [comment]
Just push another commit. Added a small test almost like what is in qp_spline_reference_line_smoother_test.cc.

**Pull request merged**